### PR TITLE
Fix race between timeout gate timing out and delegated gate returning successfully

### PR DESCRIPTION
**Pull request by charleskorn**
#### What this PR does

This PR fixes an issue that can cause store-gateways to slowly reduce the number of requests that they process, or start rejecting all requests with a `timeout waiting for concurrency gate` error if `-blocks-storage.bucket-store.max-concurrent-queue-timeout` is not 0.

There is a race in `timeoutGate.Start` between taking a token from the delegated gate, and the timeout occurring. If the delegated gate allows the request, but then the timeout happens, `timeoutGate.Start` will return an error, and so we'll never call `Done` on the delegated gate to allow another request to proceed.

If this only happens a few times, it means the store-gateway will be running fewer requests in parallel, and it's more likely a request will wait.

Eventually, if this happens enough times, the delegated gate will think all of its requests are in progress, so all requests will end up blocking and timing out after `-blocks-storage.bucket-store.max-concurrent-queue-timeout`.

#### Which issue(s) this PR fixes or relates to

#7777

#### Checklist

- [x] Tests updated.
- [n/a] Documentation added.
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [n/a] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Review by jhalterman — APPROVED** [review]
LGTM, nice test

File: pkg/storegateway/bucket_stores.go [hunk 1]
```diff
@@ -445,7 +445,7 @@ func (t timeoutGate) Start(ctx context.Context) error {
 	defer cancel()
 
 	err := t.delegate.Start(ctx)
-	if errors.Is(context.Cause(ctx), errGateTimeout) {
+	if err != nil && errors.Is(context.Cause(ctx), errGateTimeout) {
```
Comment (pracucci) [inline pkg/storegateway/bucket_stores.go:448]: Good job spotting this race! It's very nasty! Even if you have a test, I think a comment here to explain why we check `err != nil` would be very nice and useful for the future. Fancy to open a follow up PR with it?

Comment (56quarters) [reply pkg/storegateway/bucket_stores.go:448]: I can add a comment to the `timeoutGate` in #8138

Comment (charleskorn) [reply pkg/storegateway/bucket_stores.go:448]: Thanks @56quarters!

**Pull request merged**